trino/client.py

```diff
@@ -700,14 +700,13 @@ def rownumber(self) -> int:
     def __iter__(self):
         # A query only transitions to a FINISHED state when the results are fully consumed:
         # The reception of the data is acknowledged by calling the next_uri before exposing the data through dbapi.
-        while not self._query.finished or self._rows is not None:
-            next_rows = self._query.fetch() if not self._query.finished else None
+        while not self._query.finished or self._rows:
             for row in self._rows:
                 self._rownumber += 1
                 logger.debug("row %s", row)
                 yield row
 
-            self._rows = next_rows
+            self._rows = self._query.fetch() if not self._query.finished else []
```

Comment:
Through this change the above comment is not valid anymore as the `next_uri` is only acknowledged after exposing all rows from `self._rows` through dbapi. The result is the same but now the query will be hanging until the dbapi method `fetch_one` or `fetch_many` has processed all rows in the buffer.

Also in the future the fetch could be made async, done in a separate thread, which allows for continuously looping through rows and fetching the next resultset.